### PR TITLE
Maroon-152_HuygensPlateModify

### PR DESCRIPTION
**Pull request by dpdsw**
plate alignment

added scripts to plate

pre-refactor

Add SlitPlate Template

combined plate

fixed plate

added WaveGenerator objects

added sliders for plate

fixed rotation and scaling of the plate

Add plate to VR

added MLG Keys for the sliders

change position of wavegen

Added Deactivation of Wave for 0 Slit Width

WaveGenerator positioning and creation/registration

Added Shaders, Snap Plate Scaling, Number of generated WaveGenerators

moved edge generators

revertet propagation mode change

Renamed variables and functions for better readability

plate width reset

Maroon-152_HuygensPlateModify

### CODE REVIEW

File: unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+        private int CalculateGeneratorsPerSlit()
+        {        
+            int numberOfGeneratorsPerSlit = 1 + (int)((slitWidth * 10.0f)/1.5f);
+            return numberOfGeneratorsPerSlit; 
+        }
+
+        public void AddWaveGenerator()
+        {
+            var waveGenerator = new GameObject("WaveGenerator");
+            waveGenerator.transform.parent = GameObject.Find("Plate").transform;
```
Comment (michaelholly) [inline unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs:224]: Use "waveGenerator.transform.parent = gameobject.transform;"

File: unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+            waveGenerator.transform.rotation = waveGenerator.transform.parent.rotation;
```
Comment (michaelholly) [inline unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs:225]: Use "waveGenerator.transform.rotation = gameobject.transform.rotation;"

File: unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+            //waveGenerator.transform.localPosition = left.transform.localPosition;
+
+            WaveGenerator waveGeneratorScript = waveGenerator.gameObject.AddComponent<WaveGenerator>();
+            waveGeneratorScript.WaveAmplitude = 0.5f;
```
Comment (michaelholly) [inline unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs:229]: At this point, we have to consider the current amplitude, frequency, and wavelength.

File: unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+            waveGeneratorScript.WaveLength = 0.25f;
+            waveGeneratorScript.WaveFrequency = 0.5f;
+            waveGeneratorScript.SetPropagationMode(WaveGenerator.WavePropagation.Circular);
+            if (GameObject.Find("Water") != null)
```
Comment (michaelholly) [inline unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs:233]: I think we do not need this part.

File: unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+            for (int count = 0; count < totalNumberOfGenerators - 1; count++)
+            {
+                AddWaveGenerator();
+            }
+            RegisterPlateWaveGenerators();
+        }
+
+        private void RegisterPlateWaveGenerators()
+        {
+            foreach (var generator in waveGeneratorList)
```
Comment (michaelholly) [inline unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs:252]: We can simplify this to:
foreach (var generator in waveGeneratorList.Where(generator => generator != waveGenerator))

File: unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+        }
+
+        public void ResetObject()
+        {
+            throw new System.NotImplementedException();
+        }
+
+        public void ResetCubes()
+        {
+            foreach (var section in midSections){
```
Comment (michaelholly) [inline unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs:276]: Move bracket to next line

File: unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+                DestroyImmediate(section);   
+            }
+
+            ResetWaveGenerators();
+            midSections.Clear();             
+        }
+
+        public void ResetWaveGenerators()
+        {
+            foreach (var gen in waveGeneratorList)
```
Comment (michaelholly) [inline unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs:286]: We can simplify this to:
foreach (var gen in waveGeneratorList.Where(gen => gen != waveGenerator).Where(gen => gen != null))

File: unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+        }
+
+        public void AddWaveGenerator()
+        {
+            var waveGenerator = new GameObject("WaveGenerator");
+            waveGenerator.transform.parent = GameObject.Find("Plate").transform;
+            waveGenerator.transform.rotation = waveGenerator.transform.parent.rotation;
+            //waveGenerator.transform.localPosition = left.transform.localPosition;
+
+            WaveGenerator waveGeneratorScript = waveGenerator.gameObject.AddComponent<WaveGenerator>();
```
Comment (michaelholly) [inline unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs:228]: We can simplify this to:
var waveGeneratorScript = waveGenerator.AddComponent<WaveGenerator>();

File: unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+
+        private void DeactivatePlateWaveGenerators()
+        {
+            foreach (var generator in waveGeneratorList)
+                generator.GetComponent<WaveGenerator>().SetGeneratorActive(false);
+        }
+
+        public void ResetObject()
+        {
+            throw new System.NotImplementedException();
```
Comment (michaelholly) [inline unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs:271]: Please use this method for resetting the plate object.
Do not use the simple reset object script anymore.

File: unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+                    midSections[index].transform.localScale = new Vector3(scale.x, scale.y, scale.z);
+                    if(scaleInBounds)
+                    {
+                        var transformCubeSection = left.transform.localPosition.x + ((scale.x + slitWidth) * (float)(index + 1)) + scale.x / 2.0f;                                
+                        midSections[index].transform.localPosition = new Vector3(transformCubeSection, left.transform.localPosition.y, left.transform.localPosition.z);
+                    }
+                }
+        }
+
+        public void ScaleAndPositionWaveGenerators(Vector3 scale, bool scaleInBounds)
```
Comment (michaelholly) [inline unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs:170]: Hard to read.
Please simplify this method as much as possible.

File: unity/Assets/Resources/MLG/huygens_principle.xml [hunk 11]
```diff
@@ -86,6 +86,17 @@
             <English>Wave Visualization</English>
             <German>Wellenvisualisierung</German>
         </Translation>
+		
+		<Translation Key="SlitNumber">
```
[the remark this answers is not in the document]
Comment (dpdsw) [reply unity/Assets/Resources/MLG/huygens_principle.xml]: Done

File: unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+        private GameObject bottom;
+
+        [SerializeField]
+        private GameObject right;
+
+        [SerializeField]
+        private GameObject left;
+
+        [SerializeField]
+        private GameObject waveGenerator;
```
[the remark this answers is not in the document]
Comment (dpdsw) [reply unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs]: Done

File: unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+
+        [SerializeField]
+        private WaterPlane waterPlane;
```
[the remark this answers is not in the document]
Comment (dpdsw) [reply unity/Assets/0Refactored/scenes/experiments/HuygensPrinciple/scripts/SlitPlate.cs]: Done

**Pull request merged**